File: src/com/wang/Game2dEngine/action/MoveYByAction.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action;

import static java.lang.Math.abs;

/**
 * @author ricolwang
 */
public class MoveYByAction extends MoveAction
{

    double moveYBy;
    float moveYByDuration;
    double moveYBySpeed;
    double moveYByCurrent;

    public MoveYByAction()
    {
        this.moveYBy = 0;
        this.moveYByDuration = 0;
        this.moveYBySpeed = 0;
        this.moveYByCurrent = 0;
    }

    public void moveYBy(double y, float duration)
    {
        if (duration <= 0)
        {
            duration = 0;
            bImmediately = true;
        }
        this.moveYBy = y;
        this.moveYByDuration = abs(duration * 1000);
        if (!bImmediately)
        {
            this.moveYBySpeed = y / abs(duration * 1000);
        }
        this.moveYByCurrent = 0;
    }

    @Override
    public void perform(double runningTime)
    {
        if (bImmediately)
        {
            this.theSprite.setY(this.theSprite.getY() + moveYBy);
            bComplete = true;
        }

        if (bComplete)
        {
            return;
        }

        if (abs(this.moveYBy) > Action.MINIMUM)
        {
            double y = this.theSprite.getY();
            double value = this.moveYBySpeed * runningTime;
            this.moveYByCurrent += value;

            if (this.moveYBySpeed > 0)
            {
                if (this.moveYByCurrent > this.moveYBy)
                {
                    bComplete = true;
                } else
                {
                    y += value;
                }
            } else
            {
                if (this.moveYByCurrent < this.moveYBy)
                {
                    bComplete = true;
                } else
                {
                    y += value;
                }
            }

            this.theSprite.setY(y);
        } else
        {
            bComplete = true;
        }

    }
}

File: src/com/wang/Game2dEngine/action/MoveYByAction.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action

/**
 * @author ricolwang
 */
class MoveYByAction : MoveAction()
{
    var moveYBy = 0.0
    var moveYByDuration = 0f
    var moveYBySpeed = 0.0
    var moveYByCurrent = 0.0
    fun moveYBy(y: Double, duration: Float)
    {
        var duration = duration
        if (duration <= 0)
        {
            duration = 0f
            bImmediately = true
        }
        moveYBy = y
        moveYByDuration = Math.abs(duration * 1000)
        if (!bImmediately)
        {
            moveYBySpeed = y / Math.abs(duration * 1000)
        }
        moveYByCurrent = 0.0
    }

    override fun perform(runningTime: Double)
    {
        if (bImmediately)
        {
            theSprite!!.y = theSprite!!.y + moveYBy
            bComplete = true
        }
        if (bComplete)
        {
            return
        }
        if (Math.abs(moveYBy) > Action.Companion.MINIMUM)
        {
            var y = theSprite!!.y
            val value = moveYBySpeed * runningTime
            moveYByCurrent += value
            if (moveYBySpeed > 0)
            {
                if (moveYByCurrent > moveYBy)
                {
                    bComplete = true
                } else
                {
                    y += value
                }
            } else
            {
                if (moveYByCurrent < moveYBy)
                {
                    bComplete = true
                } else
                {
                    y += value
                }
            }
            theSprite!!.y = y
        } else
        {
            bComplete = true
        }
    }
}